lib/doorkeeper/openid_connect/helpers/controller.rb

```diff
@@ -6,21 +6,34 @@ module Controller
 
         def authenticate_resource_owner!
           super.tap do |owner|
-            next unless controller_path == Doorkeeper::Rails::Routes.mapping[:authorizations][:controllers] &&
-              action_name == 'new'
-            next unless pre_auth.scopes.include?('openid')
+            next unless valid_controller_path?
+            next unless scopes.include?('openid')
 
             handle_prompt_param!(owner)
             handle_max_age_param!(owner)
           end
         rescue Errors::OpenidConnectError => exception
+          handle_error(exception)
+        end
+
+        def valid_controller_path?
+          controller_path == Doorkeeper::Rails::Routes.mapping[:authorizations][:controllers] && action_name == 'new'
+        end
+
+        def scopes
+          pre_auth.scopes
+        rescue NoMethodError
```

Comment:
To be clear - this PR is obviously meant to raise questions and NOT to be merged - we're curious if we need to guarantee the validity of `pre_auth` after `resource_owner_authenticator` is run.  Are we supposed to verify that `pre_auth` is valid with a valid `client` before we get to this point?

If not, calling `scopes` can attempt to `build_scopes` in the PreAuthorization with a `nil` client.

Reply:
Thanks for looking at this @toupeira !

I might be missing something about this, so correct me if I'm wrong with my assumptions:
- `Doorkeeper::OpenIDConnect::Helpers::Controller` overrides base Doorkeeper's `authenticate_resource_owner!`, calls it (with `super`), opens the resource owner returned, and runs OIDC specific behavior (`prompt` analysis + `max_age` check)
- In order to do so, it currently requires that `pre_auth.client` is valid.  Every other part of the PreAuthorization can be missing - `redirect_uri`, `scopes`, and `response_type` can be `nil` and it'd still work fine because only `scopes` is used in this context and `scopes` in PreAuthorization does a `client` lookup if it's missing.
- Doorkeeper's `authenticate_resource_owner!` is calling doorkeeper config  `resource_owner_authenticator`

Because of that:
- Would our `resource_owner_authenticator` need to validate `pre_auth.client?` so that by the time we drop into `Doorkeeper::OpenIDConnect::Helpers::Controller:authenticate_resource_owner!` we guarantee that `pre_auth.scopes` doesn't raise an exception trying to grab `client` if `scopes` doesn't exist.

I noticed that https://github.com/doorkeeper-gem/doorkeeper-openid_connect/pull/80 puts it back in, so I think that answers the question.

That being said, I'm wondering if it's worth checking `pre_auth.valid?` rather than specifically `pre_auth.client`.  

Would it make sense to ensure that `response_type`, `client`, `redirect_uri` are valid before we take action (they seem required in the OIDC RFC?) or is it something you want to leave to the base Doorkeeper gem itself when we finish and it drops down to the `authorized_applications_controller` or `authorizations_controller`?

Let me know if it's a concern I should put on that PR.  Thanks!